sapi/fpm/fpm/fpm_main.c

```diff
@@ -1836,25 +1841,50 @@ consult the installation file that came with this distribution, or visit \n\
 
 	/* library is already initialized, now init our request */
 	request = fpm_init_request(fcgi_fd);
+	int fpm_bootstrapped = CGIG(fpm_bootstrap) && CGIG(fpm_bootstrap)[0];
 
 	zend_first_try {
-		while (EXPECTED(fcgi_accept_request(request) >= 0)) {
+		while (1) {
+			// moving this before init_request_info will break with dtrace
+			// support in php_request_startup(), can we remove?
+			if (UNEXPECTED(fpm_bootstrapped && php_request_startup() == FAILURE)) {
+				fcgi_finish_request(request, 1);
+				SG(server_context) = NULL;
+				php_module_shutdown();
+				return FPM_EXIT_SOFTWARE;
+			}
+
+			if (fpm_bootstrapped) {
+				zend_stream_init_filename(&bootstrap_file, CGIG(fpm_bootstrap));
+
+				if (zend_execute_scripts(ZEND_REQUIRE, NULL, 1, &bootstrap_file) == FAILURE) {
```

Comment:
So here is the main missing piece that I see from my quick review. You are starting potentially longer work without changing the `request_stage` (updating scoreboard). That might result into various problems with `dynamic` and `ondemand` process managers because FPM master will think that there is no work going on and consider the process as idle. It means it might not scale properly which might be problematic. It will probably cause a shorter clean up for ondemand as it bases last idle time on the end of request currently. I think it would make sense to introduce a new request stage but it probably requires a bit more thinking and carefully considering all edge cases.

Another thing to consider are reloads that usually wait for children to finish the request so it can nicely restart. The question is what should happen for bootstrap and if there's any point to wait.

Reply:
We could allow this option only for static pools, it does make some sense for dynamic, but not for ondemand i guess

Reply:
Except I guess not many users really use `static` so it could end up being a bit useless feature :) It's a bit stupid PM so I'd really like not to give it a preference. The dynamic is the default so think this one will be the most used one and from the bug reports it looks that people are using `ondemand` as well. The `dynamic` is doable so it really needs to be done.

The `ondemand` might be a bit trick but it's kind of related in some way to the reloads which should still be done cleanly. We might need to figure out how to do a proper clean up because otherwise it might end without calling `php_request_shutdown`, hanging scoreboard proc (needed for `dynamic`) and some other things.

I think the solution might be introducing event loop (`epoll`) in the worker which could hook up the signal handling (checking pipe where the signal handler writes) and then do clean shutdown if `SIGTERM` received for example. It would help with other things a well - see for example discussion in https://github.com/php/php-src/pull/4101 . I think this might be a prerequisite for this feature.